### DISCUSSION TITLE
`http.send` cache not in effect during once evaluation

### DISCUSSION THREAD

**Discussion by lukyer** [open-policy-agent/community#247] [OPA and Rego] [answered]
Hi,
during performance investigation I discovered with Wireshark that first call to `http.send` which is not cached yet is actually carried out multiple times and caching is ignored until the policy is completely evaluated. Next evaluation works OK and uses cached value instead of firing multiple same http requests.

Is it expected behaviour or some bug? I know that the policy itself is not ideal if it allows to execute `http.send` multiple times with the same parameters but it is effect caused by [this](https://github.com/open-policy-agent/community/discussions/245). Utilizing cached value during policy run would help a lot here because although the `http.send` is called multiple times it would be loaded from cache and performance impact wouldn't be so bad.

I'm trying to evaluate e.g. 1000 rules [at once during one batch evaluation](https://github.com/open-policy-agent/community/discussions/244) and these redundant `http.send` calls because of lack of caching are causing unacceptable lags when composed (tens of seconds 😄 ).

Thanks!

**Comment by anderseknert** [comment, marked as answer]
Oh, I guess this provides a clue to my [previous question](https://github.com/orgs/open-policy-agent/discussions/245#discussioncomment-3283413) about what you were trying to accomplish :) While you're right that multiple `http.send` calls in the frame of a single policy evaluation commonly is discouraged, caching should still work as expected. Could you provide a minimal example of a policy where this behavior is demonstrated?

**Reply by lukyer** [reply]
Hey, I have tried hard to reproduce this issue in minimal example but without any luck. It just works bad in our policy composed from multiple smaller policies and mainly when we used `with` keyword. Without it there is no problem with requests duplication in Wireshark. I think I will close this issue for now. If I hit it again in the future I will reopen it :) Thanks!
